service/history/historyEngine.go

```diff
@@ -1702,6 +1703,38 @@ func (e *historyEngineImpl) RecordActivityTaskStarted(
 			requestID := request.GetRequestID()
 			ai, isRunning := mutableState.GetActivityInfo(scheduleID)
 
+			// RecordActivityTaskStarted is already past scheduleToClose timeout.
+			// If at this point pending activity is still in mutable state it may be resurrected.
+			// Otherwise it would be completed or timed out already.
+			if isRunning && e.timeSource.Now().After(ai.ScheduledTime.Add(time.Duration(ai.ScheduleToCloseTimeout)*time.Second)) {
```

Comment:
What's the risk of this being a false-positive due to clock-skew or GC or something?

Reply:
> What's the risk of this being a false-positive due to clock-skew or GC or something?

This is just a trigger check. We don't want to run those checks often as they are potentially expensive with large histories. Worst case - wasted resources, increased latencies.